### PR TITLE
feat(sse): add ping_interval for keepalive pings

### LINKED ISSUE #4082: Enhancement: recurring ping in sse responce
### Summary

problem: response failed due to timeout if no events come for a long time;
proposal: add ping message every x seconds in sse responce.

### Basic Example

like in [sse-starlette](https://github.com/sysid/sse-starlette/blob/main/sse_starlette/sse.py#L122)

### Drawbacks and Impact

_No response_

### Unresolved questions

_No response_

### PR DESCRIPTION
**Pull request by TanayK07**
## Description

Add optional `ping_interval` parameter to `ServerSentEvent` that sends SSE comment keepalive pings at the specified interval to prevent connection timeouts from reverse proxies or clients.

Closes #4082

## Approach

This PR incorporates all reviewer feedback from the stalled #4098:

- **No intermediate `SSEStream` class** — overrides `to_asgi_response()` directly on `ServerSentEvent`
- **`ASGIStreamingSSEResponse` lives in `sse.py`**, not `streaming.py` — keeps SSE concerns out of generic streaming
- **Uses SSE comments for pings** (`: ping\r\n\r\n`), not `event: ping` — comments are invisible to JS `EventSource` clients
- **`anyio.Lock` on `send()`** — ping and stream tasks run concurrently, lock prevents response corruption
- **Sleep-first, then ping** — no immediate ping on connect
- **`anyio.Event` to signal shutdown** — clean task cancellation when stream ends

## Usage

```python
@get("/stream")
async def stream_handler() -> ServerSentEvent:
    async def generator():
        while True:
            data = await get_data()
            yield data

    return ServerSentEvent(generator(), ping_interval=15)
```

This is fully opt-in — `ping_interval` defaults to `None`, so existing behavior is unchanged.

## Changes

| File | Change |
|------|--------|
| `litestar/response/sse.py` | Add `ASGIStreamingSSEResponse`, add `ping_interval` to `ServerSentEvent`, override `to_asgi_response()` |
| `tests/unit/test_response/test_sse.py` | Add 5 test cases for ping behavior |
| `docs/usage/responses.rst` | Add keepalive ping documentation section |
| `docs/release-notes/changelog.rst` | Add changelog entry |

## Test Plan

- [x] SSE without `ping_interval` works unchanged (regression test)
- [x] SSE with `ping_interval` sends keepalive comments during idle periods
- [x] Pings are SSE comments (`: ping`), not events
- [x] Ping stops cleanly when stream ends (no task leaks)
- [x] Concurrent ping and data don't corrupt the response

<hr>
📚 Documentation preview 📚: https://litestar-org.github.io/litestar-docs-preview/4623

### CODE REVIEW

**Comment by Tanay10x** [comment]
Hey @provinzkraut @euri10 — friendly ping! This PR has been open for ~10 days. CI is green on all SSE tests (5527 passed) — the only failures are 3 Postgres channel pub/sub timeouts which also fail on main branch (unrelated flaky tests).

This incorporates all the review feedback from the stalled #4098:
- SSE comments for pings (invisible to EventSource clients)
- `anyio.Lock` for concurrent send safety
- No intermediate `SSEStream` class
- 100% coverage on modified lines

Would love a review when you get a chance. Happy to make any changes needed. Thanks!

**Comment by TanayK07** [comment]
Just pushed a few improvements based on self-review:

- Added input validation for `ping_interval` (rejects zero/negative values with `ImproperlyConfiguredException`)
- `to_asgi_response()` now delegates to the parent when `ping_interval` is `None`, so the non-ping code path is completely unchanged
- Added `__all__` to `sse.py` to match `streaming.py` conventions
- A few more edge case tests (empty generator, large interval)
- Docs now explicitly mention units (seconds) and typical values

CI is green across 3.9–3.13 — the one red check is a Postgres channel timeout that also happens on main (not related to SSE).

@provinzkraut @sobolevn would really appreciate a look when you have a moment — happy to adjust anything.

**Comment by TanayK07** [comment]
Update : Codecov/patch is now green (was the last remaining check). All 27 CI checks pass. This PR is ready for    
 review whenever you get a chance , Appreciato

File: litestar/response/sse.py [hunk 1]
```diff
@@ -180,3 +245,63 @@ def __init__(
...
+        Args:
+            background: Background task(s) to be executed after the response is sent.
+            cookies: A list of cookies to be set on the response.
+            headers: Additional headers to be merged with the response headers. Response headers take precedence.
+            is_head_response: Whether the response is a HEAD response.
+            media_type: Media type for the response. If ``media_type`` is already set on the response, this is ignored.
+            request: The :class:`Request <.connection.Request>` instance.
+            status_code: Status code for the response. If ``status_code`` is already set on the response, this is
+                ignored.
+            type_encoders: A dictionary of type encoders to use for encoding the response content.
```
Comment (cofin) [inline litestar/response/sse.py:277]:
```suggestion
        Args:
            request: The :class:`Request <.connection.Request>` instance.
            background: Background task(s) to be executed after the response is sent.
            cookies: A list of cookies to be set on the response.
            headers: Additional headers to be merged with the response headers. Response headers take precedence.
            is_head_response: Whether the response is a HEAD response.
            media_type: Media type for the response. If ``media_type`` is already set on the response, this is ignored.
            status_code: Status code for the response. If ``status_code`` is already set on the response, this is
                ignored.
            type_encoders: A dictionary of type encoders to use for encoding the response content.
```

Ordered `request` as the first in the docstring.

File: litestar/response/sse.py [hunk 2]
```diff
@@ -129,6 +140,53 @@ def encode(self) -> bytes:
...
+
+    async def _send(self, send: Send, payload: bytes) -> None:
+        """Send a body chunk with lock for concurrent ping/stream safety."""
+        assert self._send_lock is not None  # noqa: S101
+        async with self._send_lock:
+            await send({"type": "http.response.body", "body": payload, "more_body": True})
+
+    async def _ping(self, send: Send, stop_event: anyio.Event) -> None:
+        """Send SSE comment keepalive pings at the configured interval."""
+        assert self._ping_interval is not None  # noqa: S101
```
Comment (cofin) [inline litestar/response/sse.py]: we may want to consider a conditional with a raise instead of assert.

File: docs/release-notes/changelog.rst [hunk 3]
```diff
@@ -16,6 +16,14 @@
         applications already using Pydantic 2. Applications using Pydantic 1 or a mix
         between 1 and 2 should upgrade to Pydantic 2.
 
+    .. change:: Add ``ping_interval`` to ``ServerSentEvent`` for keepalive pings
+        :type: feature
+        :issue: 4082
```
Comment (cofin) [inline docs/release-notes/changelog.rst:22]:
```suggestion
        :type: feature
        :pr: 4623
        :issue: 4082
```

Comment (TanayK07) [reply docs/release-notes/changelog.rst:22]: my bad, updated now

File: litestar/response/sse.py (hunk 2 above)
Comment (TanayK07) [reply litestar/response/sse.py]: changed, thanks

File: litestar/response/sse.py (hunk 1 above)
Comment (TanayK07) [reply litestar/response/sse.py:277]: changed, thanks

**Comment by cofin** [comment]
@all-contributors add @TanayK07 for code, tests, docs
